spec/models/spree/stock/availability_validator_spec.rb

```diff
@@ -0,0 +1,75 @@
+require 'spec_helper'
+
+module Spree
+  module Stock
+    describe AvailabilityValidator do
+      let(:validator) { AvailabilityValidator.new({}) }
+
+      context "line item without existing inventory units" do
+        let(:order) { create(:order_with_line_items) }
+        let(:line_item) { order.line_items.first }
+
+        before do
+          expect(order.shipment.inventory_units).to be_empty
+          expect(line_item.target_shipment).to be_nil
```

Comment:
why would we put an assertion in the setup phase? if you are trying to ensure the objects are in a particular state, why not create them like that? To me, this leads to unconfident tests. It's like we don't trust our test setup and factories and make the test suite even slower.

Reply:
this is a familiar pattern to me, I am using standard app factories and before I do my testing I make sure the factories are giving me what I need for the test. so if someone changes the factories, they will need to take care of this spec. I think this is better than to force target_shipment null on the line_item because I want this spec to break if the line_item factory is changed in terms  of target_shipment. So, I want the future dev that changes the line_item.target_shipment in the generic factory to see this breaking.
same for shipment.inventory_units.

It's also avoiding having to do what the factories have already done.

Reply:
You made me think with this one. And my conclusion is that I like this pattern.
This is the _given_ part in BDD (given/when/then), it's given, doesn't necessarily need to be composed of creation statements, can be assertion. I find it really natural to have expectations in this phase.

Reply:
I don't know. It feels totally weird to me. It's the first time I see assertions in the setup phase.

> I make sure the factories are giving me what I need for the test. so if someone changes the factories, they will need to take care of this spec.

If the problem is to have valid factories, why not specific tests for them? you could even TDD while implementing the factories if you want. But why would we test their validity with every single use of them? it's enough to check it one time, not thousands. Think also how much slower our test suite would become if we applied this pattern in all tests.

> So, I want the future dev that changes the line_item.target_shipment in the generic factory to see this breaking.

If you change a generic factory tests that rely on it will break. Why the extra step?

> It's also avoiding having to do what the factories have already done.

But I wouldn't redo what factories do. If I understand you correctly, there's no point on setting up the test independently from the factory. If a particular factory sets up an object in a certain state, let's make use of it in the test. No point on redoing it.

That's what you when you first create a factory. You move the setup code you had in the test into a factory and then the test becomes simpler while making the setup reusable.

> This is the given part in BDD (given/when/then), it's given, doesn't necessarily need to be composed of creation statements, can be assertion.

But this is a plain simple unit test. The good old and arcane Setup, Exercise, Verify and Teardown phases. I feel like we're somehow reinventing things that are well established and understood by devs.

Reply:
It might as well be that I'm missing something.

Reply:
I put these lines because I think they are very important for the reader to understand the test conditions. basically, no target_shipment and no inv_units on the order.shipment
then later we have a test with no target_shipment and with inv_units.
these are things implicit in the very generic factories used order_with_line_items and then completed_order_with_totals.
I think it makes it obvious the conditions being tested.
but look, I am ok with anything actually, what's your suggestion? delete these expect statements? I can do that although I think the test wouldnt read so well.

Reply:
re reinventing things, this is simple BDD at unit test level. nothing new.
a major dimension of BDD is test readability.

Reply:
If the *given* state is important for the test, I would move the expect statements into the test where they matter. In some cases it can be written like this:
```rb
expect(object).to be_empty
add_something
expect(object).to_not be_empty

# or
expect(object).to be_empty
do_something_not_adding
expect(object).to be_empty # still
```

And that's where people invented the `change` expectation. The alternative is what Pau said about making it explicit in the setup:
```rb
object = create(:object, empty: true)
# or
object = create(:object, items: [])
```

Then you trust the factory to do the right thing and don't need to repeat it in the test. I think either is valid and it depends on the context. But I agree with Pau that the `before` block is not the natural habitat of `expect` statements. ;-)

I guess the only thing that is really important to me here and I think we all agree on that already is to **not** rely on factory defaults:

```rb
# BAD
item = create(:line_item)
item.add_stock(3)
expect(item.stock).to eq 8
# the reader has to know that the factory uses a default stock of 5
```

Reply:
ok, I will delete these lines.

>>> the underlying concern we all have is that we are not 100% confident about our factories

I trust the factories, with theses line I just wanted to communicate better the important details about them. the objective was readability.

Reply:
> Personally, I find them too big and so hard to know what they actually create for me. 

I see. These expects are a simple solution to that problem, you use the factories and assert what you need from them before your test.

Reply:
> if the factories are changed, the tests will most likely break.

That's what I understand